quickwit-indexing/src/actors/indexing_server.rs

```diff
@@ -0,0 +1,291 @@
+        self.handle.join().await
+    }
+}
+
+pub struct IndexingServer {
+    data_dir_path: PathBuf,
+    split_store_max_num_bytes: usize,
+    split_store_max_num_splits: usize,
+    metastore: Arc<dyn Metastore>,
+    storage_uri_resolver: StorageUriResolver,
+    indexing_pipeline_handles: BTreeMap<(String, String), ActorHandle<IndexingPipeline>>,
+}
+
+impl IndexingServer {
+    pub fn spawn(
+        indexer_config: IndexerConfig,
+        metastore: Arc<dyn Metastore>,
+    ) -> IndexingServerClient {
+        let universe = Universe::new();
+        let indexing_server = Self {
+            data_dir_path: indexer_config.data_dir_path,
+            split_store_max_num_bytes: indexer_config.split_store_max_num_bytes.get_bytes()
+                as usize,
+            split_store_max_num_splits: indexer_config.split_store_max_num_splits,
+            metastore,
+            storage_uri_resolver: quickwit_storage_uri_resolver().clone(),
+            indexing_pipeline_handles: Default::default(),
+        };
+        let (mailbox, handle) = universe.spawn_actor(indexing_server).spawn_async();
+        IndexingServerClient {
+            universe,
+            mailbox,
+            handle,
+        }
+    }
+
+    async fn detach_indexing_pipeline(
+        &mut self,
+        _ctx: &ActorContext<IndexingServerMessage>,
+        index_id: String,
+        source_id: String,
+    ) -> anyhow::Result<ActorHandle<IndexingPipeline>> {
+        let indexing_pipeline_handle = self
+            .indexing_pipeline_handles
+            .remove(&(index_id.clone(), source_id.clone()))
+            .with_context(|| {
+                format!(
+                    "Indexing pipeline `{}` for source `{}` does not exist.",
+                    index_id, source_id
+                )
+            })?;
+        Ok(indexing_pipeline_handle)
+    }
+
+    async fn observe_indexing_pipeline(
+        &mut self,
+        _ctx: &ActorContext<IndexingServerMessage>,
+        index_id: String,
+        source_id_opt: Option<String>,
+    ) -> anyhow::Result<HashMap<String, Observation<IndexingStatistics>>> {
+        let mut observations = HashMap::new();
+
+        if let Some(source_id) = source_id_opt {
+            let indexing_pipeline_handle = self
+                .indexing_pipeline_handles
+                .get(&(index_id.clone(), source_id.clone()))
+                .with_context(|| {
+                    format!(
+                        "Indexing pipeline `{}` for source `{}` does not exist.",
+                        index_id, source_id
+                    )
+                })?;
+            let observation = indexing_pipeline_handle.observe().await;
+            observations.insert(source_id, observation);
+        } else {
+            for ((_, source_id), indexing_pipeline_handle) in self
+                .indexing_pipeline_handles
+                .range((index_id.clone(), "".to_string())..)
```

Comment:
That's quite cryptic. 

A `HashMap<String, HashMap<String, _>>` might be less riddly.
Otherwise we can add a method called 

```fn source_ids(&self, index_id: &str) -> impl Iterator<Item=&str>```

to encapsulate that logic.